File: Lista de Exercicio Paradigmas/Lista Python/Ex4.py
class Animal:
    def som(self):
        raise NotImplementedError("Este método deve ser implementado por subclasses.")

class Cachorro(Animal):
    def som(self):
        return "Au Au!"

class Gato(Animal):
    def som(self):
        return "Miau!"

def fazer_som(animal):
    print(f'O animal faz: {animal.som()}')

cachorro = Cachorro()
gato = Gato()

fazer_som(cachorro)
fazer_som(gato)

File: Lista de Exercicio Paradigmas/Lista Java/Ex4.java
abstract class Animal {
    abstract String som();
}

class Cachorro extends Animal {
    @Override
    String som() {
        return "Au Au!";
    }
}

class Gato extends Animal {
    @Override
    String som() {
        return "Miau!";
    }
}

public class Main {
    public static void fazerSom(Animal animal) {
        System.out.println("O animal faz: " + animal.som());
    }

    public static void main(String[] args) {
        Cachorro cachorro = new Cachorro();
        Gato gato = new Gato();

        fazerSom(cachorro);
        fazerSom(gato);
    }
}
